src/coingecko/Coingecko.ts

```diff
@@ -72,8 +101,42 @@ export class Coingecko {
     currency = "usd",
     platform_id = "ethereum"
   ): Promise<[string, number]> {
-    const result: CoinGeckoPrice[] = await this.getContractPrices([contract_address], currency, platform_id);
-    return [result[0].timestamp.toString(), result[0].price];
+    const priceCache: { [addr: string]: CoinGeckoPrice } = this.getPriceCache(currency, platform_id);
+    const now: number = msToS(Date.now());
+    let tokenPrice: CoinGeckoPrice | undefined = priceCache[contract_address];
+
+    if (tokenPrice === undefined || tokenPrice.timestamp + this.maxPriceAge <= now) {
+      if (this.maxPriceAge > 0) {
+        this.logger.debug({
+          at: "Coingecko#getCurrentPriceByContract",
+          message: `Cache miss on ${platform_id}/${currency} for ${contract_address}`,
+          maxPriceAge: this.maxPriceAge,
+          tokenPrice: tokenPrice,
+        });
+      }
+
+      try {
+        // Force price cache update.
+        await this.getContractPrices([contract_address], currency, platform_id);
+      } catch (err) {
+        const errMsg = `Failed to retrieve ${platform_id}/${currency} price for ${contract_address} (${err})`;
+        this.logger.warn({
+          at: "Coingecko#getCurrentPriceByContract",
+          message: errMsg,
+        });
+        throw new Error(errMsg);
```

Comment:
I also wasn't totally sure about the best strategy to apply here, so I landed on the status quo with some additional logging. The existing implementation didn't catch any errors, so they just rippled upwards. This implementation is basically just a continuation of that, with the addition of a warning.

The impact on PD is a good point though - this would unconditionally generate a low-urgency incident, and potentially a high-urgency incident if the exception isn't caught in the upper layers. Based on that, it seems like this should probably revert to a debug message instead. Thoughts?

Reply:
OK, so in conclusion I realised it was probably an error to log the exception in getPriceByContract(), so I have relocated that logging to getContractPrices() because it's the focal point for exceptions that might be thrown. Exceptions will unconditionally be rippled upwards, but at least we guarantee that the underlying cause is always logged.

Reply:
That makes sense!